src/Analyzers/CSharp/Analyzers/UsePatternCombinators/AnalyzedPattern.cs

```diff
@@ -0,0 +1,135 @@
+
+#nullable enable
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UsePatternCombinators
+{
+    /// <summary>
+    /// Base class to represent a pattern constructed from various checks
+    /// </summary>
+    internal abstract class AnalyzedPattern
+    {
+        public readonly IOperation Target;
+
+        private AnalyzedPattern(IOperation target)
+            => Target = target;
+
+        /// <summary>
+        /// Represents a type-pattern, constructed from is-expression
+        /// </summary>
+        internal sealed class Type : AnalyzedPattern
+        {
+            public readonly TypeSyntax TypeSyntax;
+
+            public Type(TypeSyntax type, IOperation target) : base(target)
+                => TypeSyntax = type;
+        }
+
+        /// <summary>
+        /// Represents a source-pattern, constructed from C# patterns
+        /// </summary>
+        internal sealed class Source : AnalyzedPattern
+        {
+            public readonly PatternSyntax PatternSyntax;
+
+            public Source(PatternSyntax patternSyntax, IOperation target) : base(target)
+                => PatternSyntax = patternSyntax;
+        }
+
+        /// <summary>
+        /// Represents a constant-pattern, constructed from an equality check
+        /// </summary>
+        internal sealed class Constant : AnalyzedPattern
+        {
+            public readonly ExpressionSyntax ExpressionSyntax;
+
+            public Constant(ExpressionSyntax expression, IOperation target) : base(target)
+                => ExpressionSyntax = expression;
+        }
+
+        /// <summary>
+        /// Represents a relational-pattern, constructed from relational operators
+        /// </summary>
+        internal sealed class Relational : AnalyzedPattern
+        {
+            public readonly BinaryOperatorKind OperatorKind;
+            public readonly ExpressionSyntax Value;
+
+            public Relational(BinaryOperatorKind operatorKind, ExpressionSyntax value, IOperation target) : base(target)
+            {
+                OperatorKind = operatorKind;
+                Value = value;
+            }
+        }
+
+        /// <summary>
+        /// Represents an and/or pattern, constructed from a logical and/or expression.
+        /// </summary>
+        internal sealed class Binary : AnalyzedPattern
+        {
+            public readonly AnalyzedPattern Left;
+            public readonly AnalyzedPattern Right;
+            public readonly bool IsDisjunctive;
+            public readonly SyntaxToken Token;
+
+            internal Binary(AnalyzedPattern leftPattern, AnalyzedPattern rightPattern, bool isDisjunctive, SyntaxToken token, IOperation target) : base(target)
+            {
+                Left = leftPattern;
+                Right = rightPattern;
+                IsDisjunctive = isDisjunctive;
+                Token = token;
+            }
```

Comment:
should we assert anything about the syntax here?

Reply:
i.e. `SyntaxFactory.AreEquivalent(target.Syntax, rightPattern.Target.Syntax)`?

Reply:
now that we always use TryCreate this is no longer needed (the ctor is now private)